Xabe.AutoUpdater/FtpProvider.cs

```diff
@@ -0,0 +1,470 @@
+using System.Xml.Serialization;
+using Octokit;
+using FileMode = System.IO.FileMode;
+
+namespace Xabe.AutoUpdater
+{
+    /// <summary>
+    ///     Provider for FtpReleases
+    /// </summary>
+    public class FtpProvider : IReleaseProvider
+    {
+        private readonly string _appName;
+        private readonly string _ftpHosts;
+        private readonly string _ftpLogin;
+        private readonly string _ftpPassword;
+        private readonly string _ftpFolder;
+        private readonly bool _ftpPassiveMode;
+
+        /// <summary>
+        ///     Create instance of FtpProvider. 
+        /// </summary>
+        /// <param name="appName">Name of current application.</param>
+        /// <param name="pHosts">FTP HOST.</param>
+        /// <param name="pLogin">FTP Account Login</param>
+        /// <param name="pPassword">FTP Account Password</param>
+        /// <param name="pPassiveMode">FTP Passive Mode</param>
+        /// <param name="pFolder">FTP Release Folder (optional)</param>
+        public FtpProvider(string appName, string pHosts, string pLogin, string pPassword, bool pPassiveMode, string pFolder = null)
+        {
+            _appName = appName;
+            _ftpHosts = pHosts;
+            _ftpLogin = pLogin;
+            _ftpPassiveMode = pPassiveMode;
+            _ftpPassword = pPassword;
+            _ftpFolder = string.IsNullOrEmpty(pFolder) ? "/" : pFolder;
+        }
+
+        /// <inheritdoc />
+        public Task<List<string>> DownloadCurrentVersion()
+        {
+            var t = Task.Run(async () =>
+            {
+                var client = BuildClient();
+                var latestVersion = await GetLatestVersionNumber();
+                if (!string.IsNullOrEmpty(latestVersion))
+                {
+                    var remoteFile = string.Concat(_ftpFolder, $"{_appName}.{latestVersion}.zip");
+                    var tempFile = Path.GetTempFileName();
+                    if (File.Exists(tempFile))
+                        File.Delete(tempFile);
+                    client.Download(remoteFile, tempFile);
+                    if (File.Exists(tempFile))
+                    {
+                        var outputDir = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
+                        ZipFile.ExtractToDirectory(tempFile, outputDir);
+                        return Directory.GetFiles(outputDir, "*", SearchOption.AllDirectories)
+                            .ToList();
+                    }
+                }
+                return null;
+            });
+            t.Wait();
+            return t;
+        }
+
+        /// <inheritdoc />
+        public Task<string> GetLatestVersionNumber()
+        {
+            var t = Task.Run(() =>
+            {
+                string version = "0.0.0.0";
+                var client = BuildClient();
+                var remoteDirList = client.DirectoryListSimple(_ftpFolder);
+                if (remoteDirList != null && remoteDirList.Any())
```

Comment:
This is unnecessary.
DirectoryListSimple always returns array even when error occurs.
You don't have to check if it has any elements because later you use Where so even if it won't have any elements it will work.

Reply:
ok.